app_flutter/lib/state/flutter_build.dart

```diff
@@ -0,0 +1,51 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+
+import 'package:cocoon_service/protos.dart' show CommitStatus;
+
+import '../service/cocoon.dart';
+
+/// State for the Flutter Build Dashboard
+class FlutterBuildState extends ChangeNotifier {
+  /// Cocoon backend service that retrieves the data needed for this state.
+  final CocoonService _cocoonService = CocoonService();
+
+  /// How often to query the Cocoon backend for the current build state.
+  final Duration _refreshRate = Duration(seconds: 10);
+
+  /// Timer that calls [_fetchBuildStatusUpdate] on a set interval.
+  @visibleForTesting
+  Timer refreshTimer;
+
+  /// The current status of the commits loaded.
+  List<CommitStatus> statuses = [];
+
+  /// Start a fixed interval loop that fetches build state updates based on [_refreshRate].
+  void startFetchingBuildStateUpdates() async {
+    if (refreshTimer != null) {
+      // There's already an update loop, no need to make another.
+      return;
+    }
+
+    refreshTimer =
+        Timer.periodic(_refreshRate, (t) => _fetchBuildStatusUpdate());
+  }
+
+  /// Request the latest [statuses] from [CocoonService].
+  void _fetchBuildStatusUpdate() async {
+    statuses = await _cocoonService.fetchCommitStatuses();
```

Comment:
Would be nice to test that this gets called without waiting 10 seconds

Reply:
Would `WidgetTester.pump(Duration)` work for this?

Reply:
Maybe...

Reply:
https://github.com/flutter/cocoon/blob/411ae1a428db2c445da6a2dfc7ee5de5ceb5d195/app_flutter/test/state/flutter_build_test.dart#L24

I tried a basic test here, but it throws an exception since the refreshTimer is still running at the end of the test. On tear down, I call FlutterBuildState.dispose() that cancels the timer.

Reply:
I would tear down the buildState in the test and not tearDown so the test can pass then

Reply:
Okay that worked. That's an interesting issue.

The test I implemented takes less than a few seconds to run, so I think this is done.

Reply:
The widget tests should be running in a fake async zone that controls the flow of time